lib/recursive-methods.rb

```diff
@@ -1,49 +1,217 @@
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0 
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+##################################################################################################
+# Write #5
+
+# Time complexity: O(n) beacuse it runs n times - n is length of string
+# Space complexity: O(n) beacuse the functions is called n time. 
+def nested(s, first = 0, last = s.length - 1)
+  return false if s.length%2 != 0
+  return false if (first < last) && (s[first] == s[last])
+  if (first < last) && (s[first] != s[last])
+    return nested(s, first + 1, last - 1)
+  end
+  return true
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Alternative way to solve Write #5
+# Write #5
+# Time complexity: O(n) because it runs n times - n is the length of string
+# Space complexity: O(n^2) because a new string is careated when the function is called 
+# def nested(s)
+#   mid = s.length/2
+#   return true if s == ""
+#   return false if s[mid-1] == s[mid] || s[0] == s[-1]  
+#   if s.length%2 != 0 
+#     return false
+#   elsif s.length%2 == 0 && s[0] != s[-1] 
+#     return nested(s[1...-1])
+#   end
+#   return true
+# end
+
+###############################################################################################
+
+# Write #6
+# Time complexity: O(n) beacuse it runs n times - n is length of string
+# Space complexity: O(n) beacuse we don't create any new array each time the function is called. The functions are called n times - n is length of string
+def search(array, value, index = 0 )
+  return false if array == []
+  return true if array[index] == value 
+  if (index < array.length) && (array[index] != value)
+    return search(array,value,index + 1)
+  end
+  return false
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+### Alternative way to solve Write #6
+# Write #6
+# Time complexity: O(n)
+# Space complexity: O(n^2) beause a new array is careated each time the function is called
+# def search(array, value)
+#   return false if array == []
+#   return true if array[0] == value
+#   return search(array[1..-1], value)
+# end
+
+####################################################################################################
+# Write #7
+# Time complexity: O(n) because it runs n time - n is length of string
+# Space complexity: O(n) because no new array is crated when the function is called, and the functions are called n times - n is length of string
+def is_palindrome(s, first = 0, last = s.length - 1 )
+  return true if s == ""
+  return false if first < last && s[first] != s[last]
+  return true if first == last && s[first] == s[last]
```

Comment:
If first and last are the same index, then `s[first] == s[last]` by definition.

```suggestion
  return true if first == last
```